Spigot-API-Patches/0039-Added-source-to-PlayerExpChangeEvent.patch

```diff
@@ -0,0 +1,57 @@
+From 6f49e1bf0a9b07ddef4f7232fd0f5a97c6545658 Mon Sep 17 00:00:00 2001
+From: AlphaBlend <whizkid3000@hotmail.com>
+Date: Thu, 8 Sep 2016 08:47:08 -0700
+Subject: [PATCH] Added source to PlayerExpChangeEvent
+
+
+diff --git a/src/main/java/org/bukkit/event/player/PlayerExpChangeEvent.java b/src/main/java/org/bukkit/event/player/PlayerExpChangeEvent.java
+index f37491d..e856ac8 100644
+--- a/src/main/java/org/bukkit/event/player/PlayerExpChangeEvent.java
++++ b/src/main/java/org/bukkit/event/player/PlayerExpChangeEvent.java
+@@ -1,5 +1,6 @@
+ package org.bukkit.event.player;
+ 
++import org.bukkit.entity.Entity; // Paper
+ import org.bukkit.entity.Player;
+ import org.bukkit.event.HandlerList;
+ 
+@@ -8,14 +9,36 @@ import org.bukkit.event.HandlerList;
+  */
+ public class PlayerExpChangeEvent extends PlayerEvent {
+     private static final HandlerList handlers = new HandlerList();
++    // Paper start
++    private Entity source;
+     private int exp;
+ 
+     public PlayerExpChangeEvent(final Player player, final int expAmount) {
++          this(player, null, expAmount);
++    }
++
++    public PlayerExpChangeEvent(final Player player, final Entity sourceEntity, final int expAmount) {
+          super(player);
++         source = sourceEntity;
+          exp = expAmount;
+     }
+ 
+     /**
++     * Get the source that provided the experience.
++     *
++     * Be aware that this may be null depending on
++     * how the event was constructed, so it is
++     * advised to check if this is null before
++     * doing anything with it.
++     *
++     * @return The source of the experience
```

Comment:
Suggested rewording:

Returns the source of the experience gained.

@return Source of experience, or null if an entity was not the source of the experience.

May also want to outline in what conditions (ex. plugin setting experience) which this would be null or not null.

Should we maybe also consider a more general Enum specifying the cause of the Exp change?